OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs

```diff
@@ -0,0 +1,190 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Deliver produced units from outside the map.")]
+	public class ProductionByDeliveryInfo : ProductionInfo
+	{
+		[Desc("Delivering actor name. The actor needs to have either of the `Helicopter` or `Plane` traits.")]
+		[ActorReference]
+		public readonly string DeliveryActor = "c17";
+
+		[Desc("Minimal interval (in ticks) between spawning new delivery actors.")]
+		public readonly int MinimumInterval = 75;
+
+		public readonly string ReadyAudio = "Reinforce";
+
+		public override object Create(ActorInitializer init) { return new ProductionByDelivery(init, this); }
+	}
+
+	class ProductionByDelivery : Production, ITick
+	{
+		readonly Actor self;
+		readonly CPos startPos;
+		readonly CPos endPos;
+		readonly ProductionByDeliveryInfo info;
+		readonly AircraftInfo aircraft;
+		readonly bool isPlane;
+
+		int timeLeft;
+		string factionVariant;
+		List<ActorInfo> production = new List<ActorInfo>();
+
+		public ProductionByDelivery(ActorInitializer init, ProductionByDeliveryInfo info)
+			: base(init, info)
+		{
+			self = init.Self;
+
+			// Start a fixed distance away: the width of the map.
+			// This makes the production timing independent of spawnpoint
+			startPos = self.Location + new CVec(self.Owner.World.Map.Bounds.Width, 0);
+			endPos = new CPos(self.Owner.World.Map.Bounds.Left - 5, self.Location.Y);
+			this.info = info;
+
+			aircraft = self.World.Map.Rules.Actors[info.DeliveryActor].Traits.Get<AircraftInfo>();
+			isPlane = aircraft is PlaneInfo;
+		}
+
+		public void Tick(Actor self)
+		{
+			if (timeLeft-- > 0 || !production.Any())
+				return;
+
+			StartDelivery();
+			timeLeft = info.MinimumInterval;
+		}
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			production.AddRange(actorsToProduce);
+			factionVariant = raceVariant;
```

Comment:
It would be a bit more robust if you could store this together with each production item.